packages/Button/src/Button.tsx

```diff
@@ -267,61 +292,124 @@ const baseStyle = css`
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function hasCustomRoot(props: ButtonProps): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function hasLinkRoot(
+  props: LinkButtonProps | ButtonButtonProps,
+): props is LinkButtonProps {
+  return props.href != null;
+}
+
+const omitProps = [
+  'as',
+  'className',
+  'disabled',
+  'size',
+  'variant',
+  'children',
+] as const;
+
+export default class Button extends Component<ButtonProps> {
+  static displayName = 'Button';
 
   static defaultProps = {
-    variant: 'default',
-    size: 'normal',
+    variant: Variant.Default,
+    size: Size.Normal,
     className: '',
     children: null,
     disabled: false,
-    type: 'button',
   };
 
   render() {
-    const {
-      className,
-      children,
+    const { className, children, disabled, variant, size } = this.props;
+
+    const commonProps = {
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        className,
+      ),
       disabled,
-      variant,
-      size,
-      as,
-      href,
-      ...rest
-    } = this.props;
+      'aria-disabled': disabled,
+    };
 
-    let Root = href ? 'a' : 'button';
+    if (hasCustomRoot(this.props)) {
+      const Root = this.props.as;
+      const rest = omit(this.props, omitProps);
 
-    if (as) {
-      Root = as;
+      return (
+        <Root {...rest} {...commonProps}>
+          {children}
+        </Root>
+      );
+    }
+
+    if (hasLinkRoot(this.props)) {
+      const rest: HTMLElementProps<'a'> = omit(this.props, omitProps);
+      return (
+        <a {...rest} {...commonProps}>
+          {children}
+        </a>
+      );
     }
 
+    const rest: HTMLElementProps<'button'> = omit(this.props, omitProps);
+
+    // NOTE(JeT): The button's `type` will be overridden if it is in the passed-in props
     return (
-      <Root
-        {...rest}
-        href={href}
-        className={cx(
-          baseStyle,
-          buttonSizes[size],
-          buttonVariants[variant],
-          className,
-        )}
-        disabled={disabled}
-        aria-disabled={disabled}
-      >
+      <button type="button" {...rest} {...commonProps}>
         {children}
-      </Root>
+      </button>
     );
   }
 }
+
+/*
+NOTE(JeT):
+When this is defined inside the actual component, React will try to infer prop types from it,
+merging them together with ButtonProps (see LibraryManagedAttributes and MergePropTypes in @types/react).
+Unfortunately, this merging uses keyof, which appears to drop the [key: string] index signature, meaning TS won't 
+allow us to pass unrecognized props down to custom components when we're using the `as` prop. This workaround avoids
+the attempt at merging, while still getting runtime type-checking for non-TS consumers of the library.
+*/
+// @ts-ignore
+Button.propTypes = {
```

Comment:
Why the `@ts-ignore`?

Reply:
TS complains that `propTypes` doesn't exist on `Button` otherwise. I'm only adding it here (as opposed to defining it as a `static`) to avoid TS seeing it (see comment above this line for my best understanding of what's going on there), so ignoring the type error seems like it shouldn't be an issue.

Reply:
We've had issues before with setting properties on components after they're exported. Might the error you're seeing be resolved by exporting after setting the `propTypes`?

Reply:
Nope, same error. TS sees that when I defined the `Button` class I didn't have any mention of a `propTypes` static member, so I can't assign anything to it. Although... I could probably make it work by doing `static propTypes: any = { ...` and get around the issue with React inferring things that way... That feels like a better solution, I'm gonna do that.

Reply:
Done. That feels much less intrusive to me.

Reply:
🍻